Mh 252  celery setup

sharrisoncaktus opened the pull request:
This is a big one: We have a new celery task that creates calls the CloudFactory API and tracks the status of the production run that has been submitted. This includes:

* new models: CloudFactoryRun and CloudFactoryUnit
* upgraded celery to 4.2.1
* new myhpom.tasks
    * the one task to submit a CloudFactory run and store the immediate response
    * task error mail signal, which will work at present for all myhpom celery tasks.
* testing with requests-mock and fixtures/cloudfactory/*.json representing the expected inputs and responses of a cloud factory API call.

This PR requires 

* ./hsctl rebuild in order to install the new python requirements.
* python manage.py migrate to get the new models if you want to run it outside of the unittests.

If you want to hand test, try creating an AD and then using task.delay(). This will actually run the task through celery / rmq, and it will call the CloudFactory API, so be sparing.

To test error mail is currently hard with this task, because our model validations protect the data that the celery task uses, so the only way to get this task to send email is to introduce a bug in it and then register the modified task with celery (simplest way: restart hydroshare). 

So, there is a task that is devoted to poking celery and seeing that it sends email:
```
>>> from myhpom import tasks
>>> result = tasks.test_error_email.delay(data={'this': 'must be', 'serializable': 'with json'})
>>> task.status
u'FAILURE'
>>> 
```
Then go check wherever you have myhpom sending email.

sharrisoncaktus committed:
[mh-252] upgraded celery==4.2.1, everything still works (including w/irods)

sharrisoncaktus committed:
[mh-252] DocumentUrl model needed to continue working on CloudFactory integration.

Merge branch 'mh-251--document-url' into mh-252--celery-setup

* mh-251--document-url:
  [mh-251] fix migrations.
  [mh-251] test_document_url_view
  [mh-251] test_document_url_model, which revealed a bug in the model (if .key was nulled, a new key would be assigned, but it should be protected against that)
  [mh-251] renamed DocumentUrl.valid_client_ip() to .authorized_client_ip() for even greater clarity
  [mh-251] bugfix with DocumentUrl.ip_range property: the .ip field is blank (empty string) when not set, so fix the falseity test to reflect that.
  [mh-251] move the automatic setting of DocumentUrl.key and .expiration from the Manager to a pre-save receiver, so that these attributes are handled correctly in all contexts, including in the admin (which doesn't use the Manager.objects.create)
  [mh-251] create the DocumentUrlAdmin. Provide the DocumentUrl object with .url and .filename attributes to provide to the admin.
  [mh-251] rename DocumentKey to DocumentUrl, and the document_by_key view to document_url
  [mh-251] DocumentKey.expiration and .ip need to be able to be blank in admin, and .ip should be displayed as "IP" not "Ip"
  [mh-251] new document_by_key view, at /document/<KEY>/<FILENAME>
  [mh-251] rename DocumentKey.ip_in_range to the more understandable DocumentKey.valid_client_ip
  [mh-251] new DocumentKey model – holds (symmetric) keys to documents.

sharrisoncaktus committed:
[mh-252] cloudfactory models

sharrisoncaktus committed:
[mh-252] revised structure of CLOUDFACTORY_* settings

sharrisoncaktus committed:
[mh-252] added fields and data management methods to CloudFactory models

sharrisoncaktus committed:
[mh-252] draft of the celery task to submit_advancedirective_to_cloudfactory

sharrisoncaktus committed:
[mh-252] simplified and generalized the celery error email for all situations, and use it generally

sharrisoncaktus committed:
[mh-252] extract send_celery_error_mail as a separate procedure

sharrisoncaktus committed:
[mh-252] allow CFUnit.input and .output to be blank

sharrisoncaktus committed:
[mh-252] rename task, document needed tests

sharrisoncaktus committed:
Merge branch 'mh-251--document-url' into mh-252--celery-setup

* mh-251--document-url:
  [mh-251] omit filename from document url and downloaded filename
  [mh-251] "I'd suggest splitting it [this big if block] out into the conceptual checks (ip, expiration, etc) so that its clearer what conditions this view is enforcing"
  [mh-251] use require_GET
  [mh-251] "Use a callable for the default value" ... "I don't think a pre save hook is needed for either of these default field values - use default on the corresponding fields (see comment above). A check against other keys is also not needed - I think you can safely assume uuid is gonna be globally unique!"
  [mh-251] convert DocumentUrl.ip to settings.DOCUMENT_URL_IP_RANGES. "Given the use case of restricting all of these document urls from a specific range of IPs I don't see a need for each document to have its own configurable ip range. I think you should migrate this feature out of the model altogether and rely on one configurable setting."
  [mh-251] "Add advancedirective to a raw_id_filelds - there will be a lot of documents."

# Conflicts:
#	hydroshare/settings.py

sharrisoncaktus committed:
Merge branch 'mh-251--document-url' into mh-252--celery-setup

* mh-251--document-url:
  [mh-251] protect against a moved document file (I hit this when switching between iRODs and filesystem storage)

sharrisoncaktus committed:
Merge branch 'mh-251--document-url' into mh-252--celery-setup

* mh-251--document-url:
  [mh-251] DOCUMENT_URL_IP_RANGES should be empty for development work.

sharrisoncaktus committed:
[mh-252]

test_cloudfactory_models, and some adjustments to those models

sharrisoncaktus committed:
Merge commit '563d08dc13bf84bd649e170edbae644183c6f049' into mh-252--celery-setup

* commit '563d08dc13bf84bd649e170edbae644183c6f049':
  Slightly better a:hover behavior in footer
  Re-add limited test suite runner
  [mh-251] flake8 to clean up these files
  [mh-251] use @override_setting to ensure settings values
  Fix test_dashboard_view's use of etree
  Remove Scribbler dependency
  Describble dashboard.html
  Describble next_steps_no_ad_template.html
  Remove unnecessary scribbler-inserted divs
  Describble privacy.html
  Describble upload/index.html
  Describble legal.html
  Describble footer.html
  Describble why-plan.html
  Describble how-it-works.html
  Describble signup.html (NOTE: removes unused SCSS)
  Describble about.html
  [mh-266] use request.get_host() instead of the "sites" framework to get the host domain.

# Conflicts:
#	hydroshare/settings.py

sharrisoncaktus committed:
[mh-252]

fix migrations in light of scribbler deletion

sharrisoncaktus committed:
[mh-252] CloudFactory gave us their IP addresses to support, so here they are.

sharrisoncaktus committed:
[mh-252] cloud factory submit task as a class, so that we can reach in and mock methods that the class relies on.

sharrisoncaktus committed:
[mh-252] more informative __repr__ for CloudFactory models

sharrisoncaktus committed:
[mh-252] more info in error emails

sharrisoncaktus committed:
[mh-252] create the mock objects needed to test the CloudFactory celery tasks without actually hitting their API (these are unit tests, after all)

sharrisoncaktus committed:
[mh-252] no need to add another fixture .pdf for unit tests.

sharrisoncaktus committed:
[mh-252] rename cloudfactory migration

sharrisoncaktus committed:
Merge commit 'd7f9fc1de71d08487b040409b06bbafe2e0b1d04' into mh-252--celery-setup

* commit 'd7f9fc1de71d08487b040409b06bbafe2e0b1d04':
  Bail hsctl when a migrate fails.
  Restore migrations to previous name
  No, *really* fix migrations
  Fix migrations
  [MH-267] Use proper formula to determine underline color of link
  [MH-267] Set `border-bottom-color` on `.alert.alert-warning > a:hover`
  Fix disabled state button color
  Remove bottom border from upload thumbnail link
  Make left padding on alerts match left pad on dashboard
  Make left padding on .tabbed match other left pads on dashboard

sharrisoncaktus committed:
fix migrations

sharrisoncaktus committed:
Merge branch 'mh-276--fix-username' into mh-252--celery-setup

* mh-276--fix-username:
  [mh-276] fix migration
  [mh-276] fix the myhpom.models.user.get_username function to omit '=' from username, and apply this to all users in the database that have '=' in their username

sharrisoncaktus committed:
[mh-252] add an EMAIL_SUBJECT_PREFIX setting for this project

sharrisoncaktus committed:
[mh-252] bump the cloudfactory migration down in the list

sharrisoncaktus committed:
[mh-252] generalized celery error email, it works!

sharrisoncaktus committed:
[mh-252] working task with expected i/o. But interface is hard to test.

sharrisoncaktus committed:
[mh-252] tests for CloudFactory celery task completed

sharrisoncaktus committed:
[mh-252]

flake8 cleanup

sharrisoncaktus committed:
Merge branch 'mh-276--fix-username' into mh-252--celery-setup

* mh-276--fix-username:
  [mh-276]

sharrisoncaktus committed:
[mh-252]

a little better testing of the construction of the task results

sharrisoncaktus committed:
[mh-252] don't submit blank or null values to CloudFactory – better data integrity protection.

sharrisoncaktus committed:
[mh-252] a task to poke celery into sending error email.

dsummersl commented on hydroshare/settings.py:
```diff
@@ -261,12 +264,38 @@
+CLOUDFACTORY_PRODUCTION_LINES = {'TEST': 'zUyZv59Mxi'}
+CLOUDFACTORY_AUTH = {
+    'USERNAME': 'a9a8d5b3289c42c29e5229b31c971f20',
+    'PASSWORD': 'c5b2debb303a47428b0beef036a478f4',
```
These credentials need to be passed in via an environmental variable an configured in staging/production (and off for developers by default).

sharrisoncaktus replied:
Ok, good to know.

dsummersl commented on hydroshare/settings.py:
```diff
@@ -247,6 +247,9 @@
 
+# The prefix on admin emails
+EMAIL_SUBJECT_PREFIX = '[MMH] '
+
```
awesome...

dsummersl commented on hydroshare/settings.py:
```diff
@@ -261,12 +264,38 @@
+# -- CLOUDFACTORY SETTINGS -- 
+# (myhpom CloudFactory integration)
+
+CLOUDFACTORY_PRODUCTION_LINES = {'TEST': 'zUyZv59Mxi'}
```
Notice that this is only used in test cases at the moment (which makes perfect sense at the moment). Given that we're only going to have one production line how about `CLOUDFACTORY_LINE_ID` (and put whatever the actual value of this id in our secrets files as below and use an example/bogos one here so its obvious its not for real).

sharrisoncaktus replied:
Good idea.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+        ad = document_url.advancedirective
+        data = {
+            'full_name': ' '.join(
+                name
```
Use `user.get_full_name()` here?

sharrisoncaktus replied:
That wouldn't capture the middle_name. (This ^ here is another reason we should go ahead and make our own User model sooner rather than later.)

dsummersl replied:
ah, right that does make sense...CF'll probably wanna see the middle name there.

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+
+    line_id = models.CharField(
+        max_length=64, help_text="the id of the production line at CloudFactory."
+    )
```
Since we aren't having more than one line_id do we need to store it here?

dsummersl replied:
Big question here - why have two models that track the status of a CF request and response?

What we truly care about is capturing the request/response/status of a CF request associated with an AD. We won't need to query/report on individual units inside of the CF request (and anyway, we're using one unit for now as it is).

Is there some advantage to having two different models? If not I suggest we use just one model and have explicit request and response fields to track the two things we'd expect to see (and remove `message` altogether since presumably that will be captured in the response).

sharrisoncaktus replied:
The main reason for having two models is that a cloud factory run has a list of units, so it mirrors that model.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,94 @@
+    def test_submit_deleted_ad(self):
+        self.ad.delete()
+        # the post_data is moot - it won't post_run - but we still want to avoid posting to the API
+        self.task.post_run = mock_post_run({})
```
I don't think a special mocker is needed here since all you're doing is asserting that it isn't called (which should be asserted). Something like:

```
    @requests_mock.Mocker()
    def test_submit_deleted_ad(self, rm):
        self.ad.delete()
        result = self.task(*self.task_args)
        self.assertEqual(result['status'], 'ABORTED')
        self.assertIn('no longer exists', result['message'])
        self.assertFalse(rm.called)
```

dsummersl replied:
(heck you could even just use the plain jane mock library in that case)

dsummersl replied:
as mentioned above the assertion should be on the model, rather than the celery response.

sharrisoncaktus replied:
Ah, you're right! good suggestions.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,94 @@
+        self.task = CloudFactorySubmitAdvanceDirectiveRun
+        self.task_args = [self.ad.id, LINE_ID, DOCUMENT_HOST, CALLBACK_URL]
+
+    def test_submit_cloudfactory_run(self):
```
I think `mock_post_run` makes it unclear what is under test and isn't verifying what it suggests it is - the '201' asserts aren't getting executed b/c the json file doesn't have two '_' chars. I'd suggest doing away with the `post_run` method on your task as requests-mock will wire up any requests.* calls are made anyway and write tests that are a little more explicit:
```
def get_mock_response(fixture_name):
    pass  # whatever logic that is!

@requests_mock.Mocker()
class CloudFactory...TestCase...

def test_specific_404_response(self, request_mock):
     cf_response = get_mock_response('path/to/example/response.json')
     request_mock.post('url request', json=cf_response)
     # same magic wiring! and assertions.

def test_successful_response(self, request_mock):
    cf_response = get_mock_response('path/to/example/response.json')
    request_mock.post('url request', json=cf_response)
    # the meat of your current test case
     
```

Some other tests that should be here:
 * When CF returns a any non-201 code (any status code we don't expect).
 * When CF doesn't respond at all (`requests.exceptions.Timeout` although this'd be fine if it were a catchall behavior for all possible requests library exceptions - `requests.exceptions.RequestException`)

(looks like maybe you had a 404 here before from the looks of mock_post_run() but maybe it got refactored out when you went over to requests-mock?)

sharrisoncaktus replied:
Ugh, you're right about the `__` filename issue.

I *really* wanted to refactor the mock_post_run() method when I got up this morning, so I'm really glad to have this input. I'll have a go at it.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,94 @@
+                # Here we test the mapping of response_data keys to the task result
+                self.assertEqual(result['run_id'], response_data['id'])
+                self.assertEqual(result['status'], response_data['status'])
+                self.assertEqual(result['created_at'], response_data['created_at'])
```
Rather than asserting the response of the function - verify the contents of the database (`CloudFactoryRun.objects.get(...)` and the status of our CFR objects...

sharrisoncaktus replied:
Uh, yeah.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+        try:
+            ad = AdvanceDirective.objects.get(id=ad_id)
+        except:
+            cf_run.status = 'ABORTED'
```
Rather than having a magic string here, define the possible values that `status` is expected to have in the model. I would assume that on success for instance, there is a corresponding status code?

Here is an example of how we have done in the past: https://github.com/nasft/nasft-services/blob/04669864b3c9ebecc1e7a305378c1e846dd88457/mapyourshow/models.py#L762-L773

sharrisoncaktus replied:
We could use cf_run.status only to keep status responses from CloudFactory. At present, I think the choices are "Processing" and "Processed". (When you abort, the run status is still "Processing" because it does it async, and when the abort has finished it's "Processed". The Unit.status is "Aborting" after you post the abort, and "Aborted" when they have finished aborting.)

I think you're right that we should have a way to save our own internal status (such as this case, when the AD has been deleted), and maybe we shouldn't pack that into the field that holds the CloudFactory status. But that suggests another field. Unless we just preemptively delete the run object when the AD has been deleted.

dsummersl replied:
It kind of does suggest a different field, or breaking out the two uses: a way to get CF's results (a method that returns the payload that we review directly, or another field that we store some final state (success/fail) if we actually want to write queries against it), and one to store our sense of the request status (initialized/queued, aborted, errored, succeeded).

This later field is important b/c in our future ticket (254 I guess I don't another that makes sense) we need to ensure that we recover gracefully when a run fails.

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+    )
+    message = models.TextField(blank=True, help_text="Any message returned by CloudFactory.")
+    created_at = models.DateTimeField(
+        blank=True, null=True, help_text="When the run was created at CloudFactory."
```
Since this is just created when the request is made use `auto_now_add`.

sharrisoncaktus replied:
At present the .created_at field is only used to store the value from CloudFactory, not our own internal creation date. Maybe a new .inserted_at field for that?

dsummersl replied:
Part of the defensive programming (IMO anyway, YMMV for other's opinions!) is minimally parsing and storing the internal state of a third party API: if the value is important to the business use case or we need to write SQL queries against it, then parse/store it. Otherwise don't.

Here I think we care about when we made the API request (at creation of this mode instance), not when CF said that they made it.

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+    )
+
+    def __repr__(self):
+        keys = [key for key in ['id', 'line_id'] if key in self.__dict__.keys()]
```
I don't see where this is used; is __repr__ needed?

sharrisoncaktus replied:
Could probably omit. I find such things to be useful when poking around with the shell.

dsummersl replied:
Try using `__unicode__` and returning what you find useful (this is also used in the Admin list view I think so human readable == good). `__repr__` includes additional expectations on the formatting that we probably don't wanna be confined by.

sharrisoncaktus replied:
That'll be a good way to provide an admin summary of the run.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+    def post_run(self, post_data):
+        return requests.post(settings.CLOUDFACTORY_API_URL + '/runs', json=post_data)
+
+    def create_unit_input(self, document_url, document_host):
```
Since this method is intimately related with how the field in our CF run model is set, move it over into the model.

sharrisoncaktus replied:
This is the big design question on this PR: Should the CloudFactory models be tied to the needs of the AdvanceDirective, or generalized and decoupled? I designed these models to be generalized and decoupled, even though it's a _little_ more work (not much: I've already created the admin in MH-254, and it's fine). 

We could certainly tie the CloudFactory model(s) to the AD (or, the DocumentURL). I thought about doing it that way, but saw how little additional work it would be to make a generalized model, and thought that having it generalized and decoupled had some advantages.

(1) We're tracking closely to the interface provided by CloudFactory. So if they change their interface, we can more easily change these models and the tasks. The less entanglement we have with our internal models, the easier that will be. 

(2) We can use these models for other CloudFactory production tasks. Not that we envision additional uses of CloudFactory on this project, but who knows? Shane is spending a lot of money on them, and he might think of other ways to use the time he's paying for. 

(3) I'm also thinking in terms of extracting a cloudfactory_api package that can be used on other projects. Not that Caktus envisions working with CloudFactory on other projects, but who knows? If we start doing anything having to do with machine learning, it might end up making sense to employ CloudFactory to train our models (assuming a client will pay for that). (Maybe we can even sell Shane on doing some big data / ML with these Advance Directives....)

2 and 3 are pretty speculative, but I don't think 1 is – we're currently on v4 of the CloudFactory API. 

As it stands the CloudFactory models don't "know" anything about our data models. But in MH-254 I did introduce some connections in the DocumentURL model to the CloudFactory models – The DocumentURL "knows" that an associated run has the DocumentURL.url in its .input field. That seems to me to be the right direction for the dependency to flow.

dsummersl replied:
speculative, but laudable goals! If we break things out for reused I'd agree with the tighter modeling you've striven for in this PR...I'd probably go for python based generic classes rather than django based so that it'd be useful more generally.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+            run=cf_run, input=self.create_unit_input(document_url, document_host)
+        )
+        response = self.post_run(cf_run.post_data)
+        response_data = response.json()
```
catch the potential `ValueError` here when on the off chance the response is not actually json (and we'll definitely wanna store that cose in our CF run model!)

sharrisoncaktus replied:
Is this in case CloudFactory is offline or throws a 500? Maybe figure out which response.status_codes are likely to be non-json. For the status_codes that give us json, it should throw an error and email us if the response data isn't json – because it means CloudFactory has radically changed their API and we need to talk.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+                )
+            )
+
+        cf_run.__dict__.update(
```
This `model.__dict__.update` pattern doesn't feel very pythonic, and I don't think it has any advantage over a an explicit `model.field = value` pattern.

That said this does remind me that tasks should be somewhat atomic, or as atomic as we can make them.

One common pattern to prevent any potential 'partial updates' from other SQL requests is to do:

```CloudFactoryRun.objects.filter(pk=cf_run.pk).update(...all the fileds)```

This has the advantage that one query is made to both select the latest values in the database (locked to prevent others from moding that record), and updating the records we care about.

I'm not sure what 'atomic' means for this task - probably NOT wrapping the whole thing in a sql atomic block...yeah, I think it might just be the filter/update pattern I suggest above. But any thoughts on that welcome (the celery docs go into length about chaining tasks to deal so that tasks do one thing its atomic/idempotent-as-possible, etc...I don't know that we need to go that far, but that too might be a good excersize...)

sharrisoncaktus replied:
I like your solution, to do it in an atomic .update().

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+class CloudFactoryRun(models.Model):
+    """Store information about current and past CloudFactory runs."""
+
+    line_id = models.CharField(
```
It'd be helpful to have a foreign key reference to the AD that this model is related to (with on_delete=SET_NULL maybe?). That'd be helpful to see in the admin, and helpful maybe for obtaining a reference to the AD and user when constructing the request payload...

sharrisoncaktus replied:
Yes, this is the main shortcoming of my approach of generalizing the CloudFactory models. Please have a look at what I've done for the admin in MH-254 before you pass final judgment on that approach.

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+    if instance.output and (
+        isinstance(instance.output, str) or isinstance(instance.output, unicode)
+    ):
+        instance.output = json.loads(instance.output)
```
Even if we were using django 1.11 these request/response fields should be strings -- b/c we may end up dealing with non json data (in particular on a response). Rather than using signals to modify the text fields directly here, provide explicit `json_input` and `json_output` properties on the models themselves.

sharrisoncaktus replied:
I don't think the CloudFactory API ever gives non-JSON responses, FWIW. But your approach would work.

sharrisoncaktus replied:
Karen's comments on Slack:

> re: catching ValueError from response.json()...it's probably desired in general for when the other side loses its mind and starts returning html or anything non-JSON. When we don't catch the exception and make an attempt to save the raw data we get very limited information on the failure (Exception Value: No JSON object could be decoded) and need to do some re-creation to see if the other side is saying anything useful at all. https://tracker.specialtyfood.com/issues/6303 is a recent case of this, where we had to spend time to investigate "well, what IS the other side saying?". As a result of having stuff like that happen I've pretty well moved towards all api usages trying to guard against this and save at least the beginning of any non-json data that we expected to be json...

> I've gotten in habit of writing wrappers like this for everywhere we use an API that's supposed to return json: https://github.com/nasft/nasft-services/blob/develop/experient/api_support.py#L24

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,119 @@
+            - or the document it points to might have been removed.
+        - if an error occurs in this process, send support email.
+        """
+        cf_run = CloudFactoryRun.objects.create(line_id=line_id, callback_url=callback_url or "")
```
Provide an initial state that, in the worst case, gives us a sense of where we were in this task?

sharrisoncaktus replied:
like an initial status='Initialized' or some such?

dsummersl replied:
Exactly - the state of this model would then be explicitly known.

dsummersl commented on requirements.txt:
```diff
@@ -9,3 +9,5 @@ libsass==0.14.5
 bgs==0.6.1
 pyyaml==3.13
+celery==4.2.1
+requests-mock==1.5.2
```
I realize requests is captured in the 'parent' dockerfile..this might be a good time to capture our need for the requirement, and upgrade!

sharrisoncaktus replied:
Good idea. Let's go ahead and refactor out hydroshare while we're at it! :-)

dsummersl replied:
We can dream!

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,146 @@
+from django.db import models
+
+
+class CloudFactoryRun(models.Model):
```
Add an admin interface.

sharrisoncaktus replied:
I've added it in MH-254.

dsummersl reviewed (changes requested):
Sorry for the deluge of comments. Lemme see if I can summarize:
 * Use one model to capture the CF request/response state.
 * add an admin interface.
 * refactor the tests, which could be more explicit in what they do - and add more tests for corner cases.
 * Make some simple changes to the task to make it more atomic-ish.
 * document all our dependencies in requerements.txt (cause someday we're gonna get rid of the hs base!)

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-252--celery-setup

* myhpom-develop:
  Puts document urls into an AD admin area (MH-278)

sharrisoncaktus committed:
[mh-252] move CloudFactory settings to local env. "These credentials need to be passed in via an environmental variable an configured in staging/production (and off for developers by default)."

sharrisoncaktus committed:
[mh-252] (cont'd)

sharrisoncaktus committed:
[mh-252] new User.get_full_name() that includes the userdetails.middle_name.

sharrisoncaktus committed:
[mh-252] rewrite the cloudfactory models as one CloudFactoryDocumentRun model that is closely tied to the DocumentUrl model.

sharrisoncaktus committed:
[mh-252] run_id can be null, but must be unique if not null. So default must be null.

sharrisoncaktus committed:
[mh-252] add requests to our app requirements.txt.

sharrisoncaktus committed:
[mh-252] bugfix the task, no need to have separate method to make the post.

sharrisoncaktus committed:
cloudfactory_response endpoint is needed, raises 404 for now.

sharrisoncaktus committed:
[mh-252] refactored CloudFactory model tests

sharrisoncaktus committed:
[mh-252] rewrote test_cloudfactory_tasks

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-252--celery-setup

* myhpom-develop:

sharrisoncaktus commented:
I didn't add the admin to this PR – that's coming with MH-254, so I hope we can wait for it until then.

sharrisoncaktus committed:
[mh-252] Admin for CloudFactoryDocumentRun

sharrisoncaktus committed:
[mh-252] remove duplicate fixture file

dsummersl commented on myhpom/admin.py:
```diff
@@ -30,21 +32,64 @@ class StateAdmin(admin.ModelAdmin):
+            '<ul>'
+            + ''.join(
+                # Is there a better way to get a URL in the mmh-admin? I don't find a resolver.
+                '<li><a href="/mmh-admin/myhpom/cloudfactorydocumentrun/%d">%s</a></li>'
```
oh yeah, that is always a pain, finding these admin resolvers. I like django-extensions for that...`python manage.py show_urls` will then tell you that this url is...`admin:myhpom_cloudfactorydocumentrun_change`

dsummersl replied:
Getting W503 from flake8  here (don't put an operator at the beginning of a line)

sharrisoncaktus replied:
Oh, flake8, right.

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -2,145 +2,116 @@
+STATUS_PROCESSED = 'Processed'
+STATUS_VALUES = [STATUS_NEW, STATUS_DELETED, STATUS_PROCESSING, STATUS_PROCESSED]
+STATUS_MAX_LENGTH = 16
+STATUS_CHOICES = [(i, i) for i in STATUS_VALUES]
```
I like to put these into the models, as its clearer what the constants are associated with (referencing module `run.status = cloudfactory.STATUS_NEW` vs referencing model `run.status = CloudFactoryDocumentRun.STATUS_NEW`...particularly when there a module starts having multiple models in a module).

sharrisoncaktus replied:
Good idea.

dsummersl commented on myhpom/tests/test_cloudfactory_models.py:
```diff
@@ -0,0 +1,76 @@
+import os
+from django.test import TestCase
+from django.conf import settings
```
Seeing some complaints from flake8 - unused settings here, and some trailing spaces in the docstrings below.

dsummersl commented on myhpom/tests/test_cloudfactory_models.py:
```diff
@@ -0,0 +1,76 @@
+        run.save_response_content(response_content)
+        self.assertEqual(run.run_id, "SOME_RUN_ID")
+        self.assertEqual(run.status, "Processing")
+        self.assertEqual(run.created_at, parse_datetime("2018-09-24T22:27:53.000Z"))
```
Include a test for the ValueError.

sharrisoncaktus replied:
ok

sharrisoncaktus replied:
No ValueError on 201, and there's no assertNotRaises, so nothing to do here?

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,117 @@
+            run_id=data['id'] if 'id' in data else None,
+            **{key: data[key] for key in ['status', 'created_at', 'processed_at'] if key in data}
+        )
+        self.refresh_from_db()
```
I assume this refresh_from_db is only needed for the test cases - move it there so that this method does one less query under normal circumstances?

sharrisoncaktus replied:
Maybe? I'll take a look.

sharrisoncaktus replied:
It appears that the QuerySet.update() method does not refresh the current instance, so the only option (apart from the .__dict__(**...) that I like) is to assign each field individually and then .save().

dsummersl commented on myhpom/models/cloudfactory.py:
```diff
@@ -0,0 +1,117 @@
+
+import json
+import datetime
```
unused import.

sharrisoncaktus replied:
oh, flake8, right

sharrisoncaktus replied:
BTW, we didn't use flake8 early in the project, so there are a lot of other warnings. I think I'll go ahead and clean them all up in the myhpom app.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,81 @@
+
+@shared_task
+class CloudFactorySubmitDocumentRun(Task):
+    def run(self, document_url_id, document_host=None):
```
Will we ever need to pass in document_host when its supplied by a constant of the system (staging or prod)?

sharrisoncaktus replied:
Hmmm, that's a good thought.

There is currently no hostname in the settings; elsewhere we're taking it out of the request, but there is no request in this context (there will be in the context of the view that calls this task). Are we going to be adding a hostname to the settings?

dsummersl replied:
hrm, oh good point. I guess...at some point when we call this we will pull the hostname from the request?

sharrisoncaktus replied:
Yes, that's what I was thinking.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,81 @@
+        # -- no need to admin support email, this is an expected case.
+        try:
+            cf_run.document_url = DocumentUrl.objects.get(id=document_url_id)
+        except:
```
A few flake8 warnings in here about whitespace and about this bare exception - make this explicitly catch the `DocumentUrl.DoesNotExist` exception that we expect.

dsummersl commented on myhpom/tasks.py:
```diff
@@ -0,0 +1,81 @@
+
+
+@shared_task
+def test_error_email(data=None):
```
I don't see this task used anywhere - can it go?

sharrisoncaktus replied:
Sure – I just needed it while I was debugging the email.

dsummersl commented on myhpom/models/user.py:
```diff
@@ -21,6 +21,14 @@ def get_username(email):
+    name
+    for name in [self.first_name, self.userdetails.middle_name, self.last_name]
+    if name != ''  # in the not-uncommon case that one of the names is blank
+)
```
Add a test case to `UserModelTestCase` for this new implementation.

sharrisoncaktus replied:
Ok.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,89 @@
+import os
+import json
+import requests
```
flake8 complains of unused imports: requests, glob, myhpom

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,89 @@
+        run_id = CloudFactorySubmitDocumentRun(du_id)
+        cf_run = cloudfactory.CloudFactoryDocumentRun.objects.get(id=run_id)
+        self.assertEqual(cf_run.status, cloudfactory.STATUS_DELETED)
+        self.assertFalse(reqmock.called)
```
Add a test for when the request returns an exception (timeouts, etc): https://requests-mock.readthedocs.io/en/latest/response.html#raising-exceptions

sharrisoncaktus replied:
Ok.

dsummersl commented on myhpom/urls.py:
```diff
@@ -29,10 +29,14 @@
     url(r'^download/(?P<path>.*)$', myhpom_views.irods_download, name='irods_download'),
+
+    url(r'^document/cloudfactory_response/?$', 
+        myhpom_views.cloudfactory_response, name='cloudfactory_response'),
```
flake8 complaining about whitespace in here. Sigh, we should really just add flake8 to our CI pipeline :/

sharrisoncaktus replied:
Yes, but we should tone down some of the warnings a bit IMO – like whitespace at the end of a docstring or comment line.

dsummersl reviewed (changes requested):
See comments - looks pretty close!

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,89 @@
+            self.assertRaises(ValueError, self.task, self.document_url.id)
+            cf_run = self.document_url.cloudfactorydocumentrun_set.last()
+            self.assertIsNotNone(cf_run)
+            self.assertEqual(cf_run.status, cloudfactory.STATUS_NEW)
```
Little confused here - what does `STATUS_NEW` mean? I'd think if we got some error back we'd be in some kind of error state.

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,89 @@
+        self.assertRaises(ValueError, self.task, self.document_url.id)
+        cf_run = self.document_url.cloudfactorydocumentrun_set.last()
+        self.assertIsNotNone(cf_run)
+        self.assertEqual(cf_run.response_content, response_data['text'])
```
One other test case that should be in here is a non-json response from CF.

sharrisoncaktus replied:
Didn't see this – should have refreshed my page. I'll add it quick.

sharrisoncaktus committed:
[mh-252] use reverse to get admin url to CloudFactoryDocumentRun - change

sharrisoncaktus committed:
[mh-252] move CloudFactoryDocumentRun constants into the class

sharrisoncaktus committed:
[mh-252] test for new User.get_full_name()

sharrisoncaktus committed:
[mh-252] remove error email task – not needed

sharrisoncaktus committed:
[mh-252] The tedious way to (a) assign all new values in one transaction and (b) ensure that the current instance is up-to-date, while (c) minimizing database calls.

sharrisoncaktus committed:
[mh-252] added test for non-json response content

sharrisoncaktus committed:
[mh-252] added test for connection timeout

sharrisoncaktus committed:
[mh-252] code cleanup, flake8

sharrisoncaktus committed:
[mh-252] add test for non-json response in cloud factor task

sharrisoncaktus committed:
flake8

sharrisoncaktus committed:
[mh-252]

"make this explicitly catch the DocumentUrl.DoesNotExist exception that we expect."

dsummersl commented on myhpom/tests/test_user_model.py:
```diff
@@ -42,4 +39,11 @@ def test_save_invalid_user_fails(self):
+        self.assertEqual(
+            user.get_full_name(),
+            ' '.join([user.first_name, user.userdetails.middle_name, user.last_name]),
+        )
```
This assertion is a duplication of the implementation under test - An explicit 'when, then' pattern would clearly show the expected behavior (ie, `user = Userfactory(first_name='first',last_name=''); self.assertEqual('first', user.get_full_name())`)

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,100 @@
+        reqmock.post(
+            settings.CLOUDFACTORY_API_URL + '/runs', exc=requests.exceptions.ConnectTimeout
+        )
+        self.assertRaises(requests.exceptions.ConnectTimeout, self.task, self.document_url.id)
```
Add assertions on the state of the run object.

sharrisoncaktus replied:
adding a STATUS_TIMEOUT to the run class and the task

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,100 @@
+            open(os.path.join(FIXTURE_PATH, 'cloudfactory', 'post_invalid_response_201.json'), 'rb')
+        )
+        reqmock.post(settings.CLOUDFACTORY_API_URL + '/runs', **response_data)
+        self.assertRaises(ValueError, self.task, self.document_url.id)
```
Add assertions on the state of the run object to verify that its state accurately reflects the error that happened here.

sharrisoncaktus replied:
adding a STATUS_UNPROCESSABLE for both the non-json and the 422 response

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,100 @@
+        reqmock.post(settings.CLOUDFACTORY_API_URL + '/runs', **response_data)
+        self.assertRaises(ValueError, self.task, self.document_url.id)
+        cf_run = self.document_url.cloudfactorydocumentrun_set.last()
+        self.assertIsNotNone(cf_run)
```
There should be an assertion here on the value of the status code, since that is how we know handdled a 404 gracefully.

sharrisoncaktus replied:
adding STATUS_NOTFOUND

dsummersl commented on myhpom/tests/test_cloudfactory_tasks.py:
```diff
@@ -0,0 +1,100 @@
+            self.assertRaises(ValueError, self.task, self.document_url.id)
+            cf_run = self.document_url.cloudfactorydocumentrun_set.last()
+            self.assertIsNotNone(cf_run)
+            self.assertEqual(cf_run.status, CloudFactoryDocumentRun.STATUS_NEW)
```
A 422 is an error state and this won't run again...so its status should be `STATUS_PROCESSED`? (and the user would be expected to re-upload a new AD in that case?)

sharrisoncaktus replied:
422 isn't a problem with the AD – it's a problem with the input variables. So I'm using the STATUS_UNPROCESSABLE for the run status.

sharrisoncaktus committed:
[mh-252] better test for the new User.get_full_name()

sharrisoncaktus committed:
[mh-252]

new run.status values and state management, added to task and tests

sharrisoncaktus committed:
[mh-252] added 'Aborted' status for the next ticket

sharrisoncaktus committed:
[mh-252] new status choices require updating the migration.

sharrisoncaktus committed:
[mh-252] shoot the strays.

dsummersl reviewed (approved):
⭐️ ⭐️ ⭐️
